src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,92 @@
+        /// Note that DiagnosticHandler is the latest in the pipeline </param>
+        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        {
+        }
+
+        internal static bool IsEnabled()
+        {
+            return s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ResponseWriteName);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            //HttpClientHandler is responsible to call DelegatingHandler.IsEnabled() before forwarding request here.
+            //This code will not be reached if ResponseWriteName is not enabled, unless consumer unsubscribes
+            //from DiagnosticListener right after the check. So some requests happening right after subscription starts
+            //might not be instrumented. Similarly, when consumer unsubscribes, extra requests might be instumented
+
+            Guid loggingRequestId = Guid.NewGuid();
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.RequestWriteName))
+            {
+                LogHttpRequest(request,  loggingRequestId);
+            }
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+
+            LogHttpResponse(responseTask, loggingRequestId);
+            return responseTask;
+        }
+
+        #region private
+
+        private static readonly DiagnosticListener s_diagnosticListener =
+            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+
+        private static void LogHttpRequest(HttpRequestMessage request, Guid loggingRequestId)
+        {
+            long timestamp = Stopwatch.GetTimestamp();
+
+            s_diagnosticListener.Write(
+                DiagnosticsHandlerLoggingStrings.RequestWriteName,
+                new
+                {
+                    Request = request,
+                    LoggingRequestId = loggingRequestId,
+                    Timestamp = timestamp
+                }
+            );
+        }
+
+        private static void LogHttpResponse(Task<HttpResponseMessage> responseTask, Guid loggingRequestId)
+        {
+            responseTask.ContinueWith(
+                (t, s) =>
+                {
+                    long timestamp = Stopwatch.GetTimestamp();
+
+                    s_diagnosticListener.Write(
+                        DiagnosticsHandlerLoggingStrings.ResponseWriteName,
+                        new
+                        {
+                            Response = t.Status == TaskStatus.RanToCompletion ? t.Result : null,
+                            LoggingRequestId = (Guid)s,
+                            TimeStamp = timestamp,
+                            Exception = t.Exception,
+                            RequestTaskStatus = t.Status
+                        }
```

Comment:
Let me summarize: we are optimizing case when user is interested in Exceptions only.
@vancem I think it's an opportunity for the new DiagnosticSource.IdEnabled(string ,object, object) overload

```
if (diagListener.IsEnabled("Response", exception)) 
//exception could be null and user can filter out responses without exceptions
{
    diagListener.Write("Response", new {Response, exception, ...})
}

...
diagListener.Subscribe(observer, 
    (name, exception, _) => name == "Response" && exception != null});

```
User who provided name-based predicate or none at all, receive all events. 
It involves additional exception check, but otherwise, we have to call IsEnabled twice for "Exception" and "Response". It also keeps code clean and reduces number of events user has to be aware of.

Regarding common event for any exception: there are no conventions for additional context that may be provided with any exception and without context it may be useless. Activity.Current should solve this problem though.